src/pybind/mgr/volumes/fs/async_job.py

```diff
@@ -35,24 +37,39 @@ def run(self):
         while retries < JobThread.MAX_RETRIES_ON_EXCEPTION:
             vol_job = None
             try:
+                if not self.async_job.run_event.is_set():
+                    log.info('will wait for run_event to bet set. postponing '
+                             'getting jobs until then')
+                    self.async_job.run_event.wait()
+                    log.info('run_event has been set, waiting is complete. '
+                             'proceeding to get jobs now')
+
                 # fetch next job to execute
                 with lock_timeout_log(self.async_job.lock):
                     while True:
                         if self.should_reconfigure_num_threads():
                             log.info("thread [{0}] terminating due to reconfigure".format(thread_name))
                             self.async_job.threads.remove(self)
                             return
+
                         timo = self.async_job.wakeup_timeout
                         if timo is not None:
                             volnames = list_volumes(self.vc.mgr)
                             missing = set(volnames) - set(self.async_job.q)
                             for m in missing:
                                 self.async_job.jobs[m] = []
                                 self.async_job.q.append(m)
+
+                        if not self.async_job.run_event.is_set():
```

Comment:
Why is this required?

Reply:
This is located in a `while` loop that'll run infinitely until a job is found. And since this loop runs without acquiring a lock. Since we must to wait without acquiring the lock, this code is breaking out of the loop.

Reply:
Otherwise thread won't stop until this iteration is finished which means `get_job()`, `register_job()`, `execute_job()` and `unregister_job()` will be run unnecessarily.

Reply:
Marking this comment resolved since it was discussed and changes have moved forward since.